nixos/modules/services/hardware/logiops.nix

```
@@ -0,0 +1,164 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) mdDoc mkEnableOption mkIf mkOption types;
+  cfg = config.services.keyd;
+in {
+  options.services.logiops = {
+    enable = mkEnableOption (mdDoc
+      "logiops, an unofficial userspace driver for HID++ Logitech devices");
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      example = ''
+        devices: ({
+          name: "Wireless Mouse MX Master";
+          smartshift:
+          {
+            on: true;
+            threshold: 30;
+            torque: 50;
+          };
+          hiresscroll:
+          {
+            hires: true;
+            invert: false;
+            target: false;
+          };
+          dpi: 1000;
+
+          buttons: (
+            {
+              cid: 0xc3;
+              action =
+              {
+                type: "Gestures";
+                gestures: (
+                  {
+                    direction: "Up";
+                    mode: "OnRelease";
+                    action =
+                    {
+                      type: "Keypress";
+                      keys: ["KEY_UP"];
+                    };
+                  },
+                  {
+                    direction: "Down";
+                    mode: "OnRelease";
+                    action =
+                    {
+                      type: "Keypress";
+                      keys: ["KEY_DOWN"];
+                    };
+                  },
+                  {
+                    direction: "Left";
+                    mode: "OnRelease";
+                    action =
+                    {
+                      type: "CycleDPI";
+                      dpis: [400, 600, 800, 1000, 1200, 1400, 1600];
+                    };
+                  },
+                  {
+                    direction: "Right";
+                    mode: "OnRelease";
+                    action =
+                    {
+                      type = "ToggleSmartshift";
+                    }
+                  },
+                  {
+                    direction: "None"
+                    mode: "NoPress"
+                  }
+                );
+              };
+            },
+            {
+              cid: 0xc4;
+              action =
+              {
+                type: "Keypress";
+                keys: ["KEY_A"];
+              };
+            }
+          );
+        });
+      '';
+      description = mdDoc ''
+        Extra configuration that is appended to the end of the file.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.etc."logid.cfg".text = ''
+      ${cfg.extraConfig}
+    '';
```

The last PR was halted due to work on the libconfig generator (#208747, #246115). The generator is now complete, with a type and lib and all, so let's use that instead.

```suggestion
  cfg = config.services.keyd;
  format = pkgs.formats.logiops { };
in {
  options.services.logiops = {
    enable = mkEnableOption "logiops, an unofficial userspace driver for HID++ Logitech devices";

    settings = mkOption {
      type = types.submodule {
        freeformType = format.type;
      };
      description = ''
        Configuration for logiops.
        
        See <https://github.com/PixlOne/logiops/wiki/Configuration> for more details.
        Also see <https://github.com/PixlOne/logiops/blob/main/logid.example.cfg> for an examle config.
      '';
      default = { };
    };
  };

  config = mkIf cfg.enable {
    environment.etc."logid.cfg".text = format.generate "logid.cfg" cfg.settings;
```

---

I'm currently in the process of implementing this; am I right in assuming I would need to add a new `logiops` key to the `pkgs/pkgs-lib/formats.nix` file?

---

No, it's already implemented at `pkgs/pkgs-lib/formats/libconfig`. The diff I provided above *should* work, but there might be other issues to clear up.

---

We could try to rewrite the example as a freeform expression as well